# PAL-seq
This script has two modes:
> Mode 1:
1. a text string which defines the prefix name of the output files (-n, required)
2. a text string which specifies the output directory (-d, optional)
3. a reference file with annotation in bed format (-r, required)
4. a STAR-aligned bam file (-b, required)
5. the un-trimmed fastq file from read1 (-f1, required)
6. the fastq file from read2 (-f2, optional, only if 'check_pa_tail' is 'True')
7. intensity file from read2 (-i, required)
8. poly(A) site annotation in bed format (-p, optional)

> Mode 2:
1. a text string which defines the prefix name of the output files (-n, required)
2. a text string which specifies the output directory (-d, optional)
3. T_signal output file generated by this script (-s, required)

Note: When using LSF, make sure to add -n 20 for multiprocessing
It calculates poly(A) tail length based on a Gaussian (mixture) hidden markov model trained on a subset of the data
It outputs these files:
1. all poly(A) tags
2. median poly(A) tail length and total number of tags for each gene
3. mean poly(A) tail length and total number of tags for each gene
4. HMM model 
5. A temporary file containing converted T-signal will be written out on the disk to save memory usage
	and it can be deleted in the end.
6. A temporary file containing the HMM states of each read cycle for each cluster, which can be deleted manually.
